### DISCUSSION TITLE
RFC: Timer API

### DISCUSSION THREAD

**Discussion by rleh** [modm-io/modm#701] [Ideas]
Currently the only Timer driver(s) in modm are written for STM32 and the API is very hardware-specific. See e.g. https://docs.modm.io/develop/api/nucleo-g474re/classmodm_1_1platform_1_1Timer8.html

I propose to rename the current API to TimerHal (similar to SpiHal) and to add APIs for different functionalities (PWM generation, encoder input, DC- and BLDC- motors, ...)

- ...

---

cc @TomSaw @salkinium

**Comment by TomSaw** [comment]
I suggest two clearly separated, stacked APIs:

## Low lvl Timer API - Feature complete
Gives unified, generic access to timers, independent of the MCU nor the MCU-family.
CTC, InputCapture, PWM single-slope, PWM dual-slope, Encoder-input, ...

## High lvl Timer API - Application oriented
Inherits or aggregates Low lvl API and feels like a peripheral driver. Like with peripheral drivers, the application-designer don't wants to be confronted with the inner mechanisms and topologies. He/She/It just wants the functionality that makes use of Timers.

#### Examples
- Instead of exposing _Input Capture_ there are methods like **measureFrequency...** 
- _Clear Timer Compare CTC_ will be configured with methods **setFrequency...**
- Facade different PWM-Modes (single-slope, dual-slope, ...) behind a unified set of methods. `Not sure with this. It may be too much -> the border of simplification`
